src/compiler.ts

```diff
@@ -7252,6 +7261,8 @@ export class Compiler extends DiagnosticEmitter {
     newExpr: ExpressionRef,
     /** Old value being replaced. */
     oldExpr: ExpressionRef,
+    /** The type shared between expressions. */
+    type: Type,
```

Comment:
According to my comments above, this would then become `newExpr, newType, oldExpr, oldType`, with `newType` used on `__retain` and `oldType` used on `__release`.